tm2/pkg/p2p/discovery/discovery.go

```diff
@@ -0,0 +1,243 @@
+package discovery
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/gnolang/gno/tm2/pkg/amino"
+	"github.com/gnolang/gno/tm2/pkg/p2p"
+	"github.com/gnolang/gno/tm2/pkg/p2p/conn"
+	"github.com/gnolang/gno/tm2/pkg/p2p/types"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	// Channel is the unique channel for the peer discovery protocol
+	Channel = byte(0x50)
+
+	// discoveryInterval is the peer discovery interval, for random peers
+	discoveryInterval = time.Second * 3
+
+	// maxPeersShared is the maximum number of peers shared in the discovery request
+	maxPeersShared = 30
+)
+
+// descriptor is the constant peer discovery protocol descriptor
+var descriptor = &conn.ChannelDescriptor{
+	ID:                  Channel,
+	Priority:            1,       // peer discovery is high priority
+	SendQueueCapacity:   20,      // more than enough active conns
+	RecvMessageCapacity: 5242880, // 5MB
+}
+
+// Reactor wraps the logic for the peer exchange protocol
+type Reactor struct {
+	// This embed and the usage of "services"
+	// like the peer discovery reactor highlight the
+	// flipped design of the p2p package.
+	// The peer exchange service needs to be instantiated _outside_
+	// the p2p module, because of this flipped design.
+	// Peers communicate with each other through Reactor channels,
+	// which are instantiated outside the p2p module
+	p2p.BaseReactor
+
+	ctx      context.Context
+	cancelFn context.CancelFunc
+
+	discoveryInterval time.Duration
+}
+
+// NewReactor creates a new peer discovery reactor
+func NewReactor(opts ...Option) *Reactor {
+	ctx, cancelFn := context.WithCancel(context.Background())
+
+	r := &Reactor{
+		ctx:               ctx,
+		cancelFn:          cancelFn,
+		discoveryInterval: discoveryInterval,
+	}
+
+	r.BaseReactor = *p2p.NewBaseReactor("Reactor", r)
+
+	// Apply the options
+	for _, opt := range opts {
+		opt(r)
+	}
+
+	return r
+}
+
+// OnStart runs the peer discovery protocol
+func (r *Reactor) OnStart() error {
+	go func() {
+		ticker := time.NewTicker(r.discoveryInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-r.ctx.Done():
+				r.Logger.Debug("discovery service stopped")
+
+				return
+			case <-ticker.C:
+				// Run the discovery protocol //
+
+				// Grab a random peer, and engage
+				// them for peer discovery
+				peers := r.Switch.Peers().List()
```

Comment:
How many elements peers are we expecting to have here? are they always on memory? I thing that this interface should return an iterator instead.